File: scripts/decode-qr.py
#!/usr/bin/env python3
"""Decode QR codes from image files using pyzbar (same lib class as real scanners).

Usage: python3 scripts/decode-qr.py <image1> [image2 ...]
Prints one line per image: PATH\tOK|FAIL\t<decoded payloads>
Exit code 0 only if every image decoded to at least one QR.
"""
import sys
from PIL import Image
from pyzbar.pyzbar import decode, ZBarSymbol

ok_all = True
for path in sys.argv[1:]:
    try:
        img = Image.open(path).convert("L")
    except Exception as e:  # noqa: BLE001
        print(f"{path}\tERROR\t{e}")
        ok_all = False
        continue
    results = decode(img, symbols=[ZBarSymbol.QRCODE])
    if results:
        payloads = ", ".join(r.data.decode("utf-8", "replace") for r in results)
        print(f"{path}\tOK\t{payloads}")
    else:
        print(f"{path}\tFAIL\t<no QR detected>")
        ok_all = False

sys.exit(0 if ok_all else 1)

File: scripts/decode-qr.mjs
/**
 * Decode QR codes from image files using the pure-JS jsQR decoder (vendored).
 * Usage: node scripts/decode-qr.mjs <image1> [image2 ...]
 * Prints: PATH  OK|FAIL  <payload>
 * Exit 0 only if every image decoded.
 */
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";
import sharp from "sharp";

const require = createRequire(import.meta.url);
const __dirname = dirname(fileURLToPath(import.meta.url));
const jsQR = require(resolve(__dirname, "vendor/jsQR.js"));

let okAll = true;
for (const path of process.argv.slice(2)) {
  try {
    const { data, info } = await sharp(path)
      .ensureAlpha()
      .raw()
      .toBuffer({ resolveWithObject: true });
    const result = jsQR(
      new Uint8ClampedArray(data.buffer, data.byteOffset, data.length),
      info.width,
      info.height
    );
    if (result && result.data) {
      console.log(`${path}\tOK\t${result.data}`);
    } else {
      console.log(`${path}\tFAIL\t<no QR detected>`);
      okAll = false;
    }
  } catch (e) {
    console.log(`${path}\tERROR\t${e.message}`);
    okAll = false;
  }
}
process.exit(okAll ? 0 : 1);
